### PR TITLE
Unique

### PR DESCRIPTION
**Pull request by baarkerlounger**
https://github.com/SciRuby/daru/issues/419

### CODE REVIEW

File: lib/daru/dataframe.rb [hunk 1]
```diff
@@ -705,6 +705,14 @@ def rolling_fillna(direction=:forward)
       dup.rolling_fillna!(direction)
     end
 
+    # Return unique rows by vector specified or all vectors
+    def unique(vecs=[])
+      vectors = vecs.empty? ? self.vectors.map(&:to_s) : vecs
+      grouped = self.group_by(vectors)
+      indexes = grouped.groups.values.map { |v| v[0] }
+      self.row[*indexes]
```
Comment (zverok) [inline lib/daru/dataframe.rb]: Unnecessary `self.`s

File: lib/daru/dataframe.rb [hunk 2]
```diff
@@ -705,6 +705,14 @@ def rolling_fillna(direction=:forward)
       dup.rolling_fillna!(direction)
     end
 
+    # Return unique rows by vector specified or all vectors
```
Comment (zverok) [inline lib/daru/dataframe.rb]: Docs should be more detailed, with some examples and description how uniqueness is considered. And param and return value descriptions.

File: spec/dataframe_spec.rb [hunk 3]
```diff
@@ -1820,6 +1820,27 @@
...
+      before { subject.unique }
+      its(:'a.to_a') { is_expected.to eq [1, 2, 3, 4, 5] }
+      its(:'b.to_a') { is_expected.to eq ['a', 'b', 'c', 'd', 'c'] }
+    end
+
+    context 'given a vector' do
+      before { subject.unique(:b) }
+      its(:'a.to_a') { is_expected.to eq [1, 2, 3, 4] }
+      its(:'b.to_a') { is_expected.to eq ['a', 'b', 'c', 'd'] }
+    end
```
Comment (zverok) [inline spec/dataframe_spec.rb]: This spec contradicts method definition, which implies array of vector names should be passed.

File: lib/daru/dataframe.rb [hunk 4]
```diff
@@ -705,6 +705,14 @@ def rolling_fillna(direction=:forward)
...
+    def unique(vecs=[])
```
Comment (zverok) [inline lib/daru/dataframe.rb]: a) method should be named `uniq` (see Ruby's Array API: https://docs.ruby-lang.org/en/2.4.0/Array.html#method-i-uniq)
b) It probably should also support blocks, like `Array#uniq`

**Comment by baarkerlounger** [comment]
Requires https://github.com/SciRuby/daru/pull/420 to be merged first as the tuple sorting fix is needed but shouldn't form part of this PR.

File: spec/dataframe_spec.rb [hunk 5]
```diff
@@ -1820,6 +1820,33 @@
...
+    context 'given a vector' do
+      it do
+        result = df.uniq("color")
+        expect(result.shape.first).to eq 2
+      end
+    end
+
+    context 'given an array of vectors' do
+      it do
+        result = df.uniq(["color", "director_name"])
```
Comment (zverok) [inline spec/dataframe_spec.rb]: Why not `uniq(:color, :director_name)` (e.g. splatting arguments)? It looks more reasonable for me.

Comment (baarkerlounger) [reply spec/dataframe_spec.rb]: Mostly just because the `group_by` takes an array but both work for me. Have changed.

**Comment by baarkerlounger** [comment]
@zverok anything left on this?

**Pull request merged**